### PR TITLE
refactor: change return error.respond() to raise

### LINKED ISSUE #6665: Register error handler for custom errors and change return statements to raise
https://github.com/fossasia/open-event-server/blob/9ba29f50ff2907e911dba2a51661015183ca658a/app/api/helpers/errors.py#L62-L75

https://flask.palletsprojects.com/en/1.1.x/errorhandling/#generic-exception-handlers

### PR DESCRIPTION
**Pull request by Rits1272**
Fixes #6665 

#### Checklist

- [x] I have read the [Contribution & Best practices Guide](https://blog.fossasia.org/open-source-developer-guide-and-best-practices-at-fossasia) and my PR follows them.
- [x] My branch is up-to-date with the Upstream `development` branch.
- [x] The unit tests pass locally with my changes 
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added necessary documentation (if appropriate)

- [x] All the functions created/modified in this PR contain relevant docstrings.

### CODE REVIEW

File: app/instance.py [hunk 1]
```diff
@@ -257,13 +258,13 @@ def internal_server_error(error):
...
-def ratelimit_handler(error):
+@app.errorhandler(HTTPException)
+def error_handler(error):
     return make_response(json.dumps({
-            'status': 429,
-            'title': 'Request Limit Exceeded'
+            'status': error.code,
+            'title': error.name
         }),
-        429,
```
Comment (iamareebjamal) [inline app/instance.py]: Why are you changing this?

File: app/instance.py [hunk 2]
```diff
@@ -269,5 +270,17 @@ def ratelimit_handler(error):
         })
 
 
+@app.errorhandler(HTTPException)
```
Comment (iamareebjamal) [inline app/instance.py]: We don't want to handle any HTTPException, we want to handle our custom exceptions

**Comment by Rits1272** [comment]
Is that what we want @iamareebjamal?

File: app/instance.py [hunk 3]
```diff
@@ -269,5 +269,53 @@ def ratelimit_handler(error):
...
+@app.errorhandler(403)
+def forbidden_error_handler(error):
+    return make_response(json.dumps({
+            'status': 403,
+            'title': 'Unprocessable Entity'
+        }),
+        403,
+        {
+            'Content-Type': 'application/vnd.api+json'
+        })
```
Comment (iamareebjamal) [inline app/instance.py]: Still wrong

Comment (Rits1272) [reply app/instance.py]: Can you kindly please tell me @iamareebjamal what changes I need to do? I am registering error handler just as I did before for 429.

Comment (iamareebjamal) [reply app/instance.py]: You can read the issue again

Comment (Rits1272) [reply app/instance.py]: Okay

**Comment by Rits1272** [comment]
Please review my changes @iamareebjamal

**Comment by iamareebjamal** [comment]
It's still wrong. Read the issue again

**Comment by Rits1272** [comment]
Please again check @iamareebjamal. Sorry for bothering you

**Comment by iamareebjamal** [comment]
Still wrong

**Comment by Rits1272** [comment]
Please check @iamareebjamal

**Comment by iamareebjamal** [comment]
You won't be able to fix the issue unless you understand why we are doing it

**Comment by Rits1272** [comment]
I have reread the issue several times. Is that what we want to achieve @iamareebjamal ?

File: app/instance.py [hunk 4]
```diff
@@ -280,5 +282,17 @@ def ratelimit_handler(error):
     )
 
 
+@app.errorhandler(UnprocessableEntity)
+@app.errorhandler(MethodNotAllowed)
+@app.errorhandler(ForbiddenException)
+@app.errorhandler(ConflictException)
```
Comment (iamareebjamal) [inline app/instance.py]: We want to use generic error handler, not list every exception

File: app/instance.py [hunk 5]
```diff
@@ -280,5 +282,17 @@ def ratelimit_handler(error):
...
+def handle_exception(error):
+    return make_response(
+        json.dumps(jsonapi_errors([error.to_dict()])),
+        error.status,
+        {'Content-Type': 'application/vnd.api+json'},
+    )
```
Comment (iamareebjamal) [inline app/instance.py]: This changes specific errors to single undiscernable error

**Comment by Rits1272** [comment]
Hey @iamareebjamal, I removed all the listed exceptions. Please check if I have made the desired changes

File: app/instance.py [hunk 6]
```diff
@@ -280,5 +280,14 @@ def ratelimit_handler(error):
     )
 
 
+@app.errorhandler(Exception)
```
Comment (iamareebjamal) [inline app/instance.py]: We are back to square one, now it'll catch every exception! The issue clearly says add a handler for custom exceptions defined by the app

**Comment by Rits1272** [comment]
Please check now @iamareebjamal

File: app/instance.py [hunk 7]
```diff
@@ -280,5 +280,14 @@ def ratelimit_handler(error):
     )
 
 
+@app.errorhandler(JsonApiException)
+def handle_exception(error):
+    return make_response(
+        json.dumps({'status': error.status, 'title': error.title}),
+        error.status,
+        {'Content-Type': 'application/vnd.api+json'},
+    )
```
Comment (iamareebjamal) [inline app/instance.py]: OK, finally. Now please return the response using respond method

**Comment by Rits1272** [comment]
I have made the changes. Please check @iamareebjamal

File: app/instance.py [hunk 8]
```diff
@@ -280,5 +281,11 @@ def ratelimit_handler(error):
     )
 
 
+@app.errorhandler(JsonApiException)
+def handle_exception(error):
+    return ErrorResponse({'source': ''}, status=error.status,
+                         detail=error.detail, title=error.title).respond()
```
Comment (iamareebjamal) [inline app/instance.py]: Nope

**Comment by Rits1272** [comment]
Please review the changes @iamareebjamal. Is it the same respond method you were talking about?

**Comment by iamareebjamal** [comment]
no

**Comment by Rits1272** [comment]
Kindly check this out @iamareebjamal

**Comment by iamareebjamal** [comment]
As I said before, you have to use the respond method

**Comment by Rits1272** [comment]
Hello @iamareebjamal. Can you please guide me with respond method? I am guessing that it is the respond method from the ErrorResponse class.

**Comment by iamareebjamal** [comment]
Yes

**Comment by niranjan94** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 2
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/fossasia/open-event-orga-server/pullRequest?prid=4876865&bid=16162998)

File: app/instance.py [hunk 9]
```diff
@@ -9,7 +9,7 @@
 import stripe
 from celery.signals import after_task_publish
 from envparse import env
-from flask import Flask, json, make_response
+from flask import Flask, json, make_response, jsonify
```
Comment (niranjan94) [inline app/instance.py]: Codacy Issue found: [Unused jsonify imported from flask](https://app.codacy.com/manual/fossasia/open-event-orga-server/pullRequest?prid=4876865)

File: app/instance.py [hunk 10]
```diff
@@ -280,5 +281,11 @@ def ratelimit_handler(error):
     )
 
 
+@app.errorhandler(JsonApiException)
+def handle_exception(error):
+    return ErrorResponse({'source': error.source}, status=error.status,
```
Comment (niranjan94) [inline app/instance.py]: Codacy Issue found: [Trailing whitespace](https://app.codacy.com/manual/fossasia/open-event-orga-server/pullRequest?prid=4876865)

**Comment by Rits1272** [comment]
Please check @iamareebjamal. Kindly tell if I am doing something wrong

**Comment by iamareebjamal** [comment]
Yes it's still incorrect

**Comment by Rits1272** [comment]
Hello @iamareebjamal  I am using the respond method as you said earlier. Can you please explain a little what I am doing wrong?

**Comment by iamareebjamal** [comment]
You are recreating the object when you already have the object

**Comment by Rits1272** [comment]
Hello @iamareebjamal. I am in a confusion. The error object that we will be getting from the handler will be of an **exception** class, then how we will use the respond method of **ErrorResponse** class? So I used the **to_dict** method of the JsonApiException class. Kindly check

**Comment by iamareebjamal** [comment]
The errors you channged from return to raise are of ErrorResponse class which have response method. They don't even inherit from JsonApiException, so I don't know what the confusion is

The issue was as simple as instead of return error.response, catch the error in exception handler and return the response there by calling respond method, but you did everything except that including catching status codes, generic exceptions and whatnot. It is not really confusing, you are definitely making it up to be

**Comment by iamareebjamal** [comment]
When the previous behaviour of the application was that it returned the response generated from the respond method of error response class, how will it ever work on catching JsonApiException and returning `to_dict`? Do you see any problem in that?

**Comment by Rits1272** [comment]
Thank you for the detailed explanation @iamareebjamal. I know value of your time. Sorry for the misunderstanding on my part. Please check the changes.

**Comment by iamareebjamal** [comment]
Still wrong. Read my 2 comments carefully

**Comment by Rits1272** [comment]
Please check this @iamareebjamal !

File: app/api/helpers/exceptions.py [hunk 11]
```diff
@@ -1,7 +1,8 @@
 from flask_rest_jsonapi.exceptions import JsonApiException
+from app.api.helpers.errors import ErrorResponse
 
 
-class UnprocessableEntity(JsonApiException):
+class UnprocessableEntity(JsonApiException, ErrorResponse):
```
Comment (iamareebjamal) [inline app/api/helpers/exceptions.py]: Remove ErrorResponse

**Comment by Rits1272** [comment]
Please review again @iamareebjamal

**Comment by Rits1272** [comment]
Kindly please check @iamareebjamal

**Comment by Rits1272** [comment]
I have made the required changes @iamareebjamal

**Comment by niranjan94** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones removed
==============
+ app/api/helpers/permission_manager.py  -2
+ app/api/role_invites.py  -2
+ app/api/event_invoices.py  -1
+ app/api/orders.py  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/fossasia/open-event-orga-server/pullRequest?prid=4876865&bid=16162998)

**Comment by niranjan94** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones removed
==============
+ app/api/event_invoices.py  -1
+ app/api/role_invites.py  -2
+ app/api/helpers/permission_manager.py  -2
+ app/api/orders.py  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/fossasia/open-event-orga-server/pullRequest?prid=4876865&bid=16162998)

**Comment by niranjan94** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 2
           

Clones removed
==============
+ app/api/event_invoices.py  -1
+ app/api/helpers/permission_manager.py  -2
+ app/api/orders.py  -1
+ app/api/role_invites.py  -2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/fossasia/open-event-orga-server/pullRequest?prid=4876865&bid=16162998)

**Comment by iamareebjamal** [comment]
Finally, after 48 commits for a small change

**Pull request merged**